File: Assignment1/PS.java
/**
 * This class contains the priority scheduling algorithm
 * @author Eddy Qiang - CPSC 449 - Tutorial T03
 * @version 1.2
 * @since September 20, 2018
 */
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Scanner;

public class PS
{
	PrintWriter writer;
	ArrayList<Integer> PS_input;
	ArrayList<Integer> burstTime = new ArrayList<Integer>();
	ArrayList<Integer> priority = new ArrayList<Integer>();
	HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();	
	int runningTime = 0;
	int count = 0;
	double totalTime = 0;

	public PS(String inputFileName, String outputFileName) throws IOException
	{
		PS_input = new fileReader().readInput(inputFileName);
		writer = new PrintWriter(new BufferedWriter(new FileWriter(outputFileName, false)));
	}
	
	public void solve() 
	{
		for(int i = 0; i < PS_input.size(); i++)
		{
			burstTime.add(PS_input.get(i));
			i++;
			priority.add(PS_input.get(i));
		}
		
		for(int i = 0; i < burstTime.size(); i++)
		{
			map.put(priority.get(i), burstTime.get(i));	
		}
		Collections.sort(priority);
		
		for(int i = 0; i < priority.size(); i++)
		{
			writer.println(runningTime);
			totalTime += runningTime;
			runningTime += map.get(priority.get(i));
			count += 1;
		}
		double averageWaitTime = totalTime/count;
		writer.println(averageWaitTime);
		writer.close();
	}
	
	public static void main(String[] args) throws IOException 
	{
		String inputFileName = args[0];
		String outputFileName = args[1];
		PS ques3 = new PS(inputFileName, outputFileName);
		ques3.solve();
	}

	private class fileReader
	{
		public ArrayList<Integer> readInput(String fileName)
		{
			ArrayList<Integer> input = new ArrayList<Integer>();
			File file = new File(fileName);
			Scanner sc = null;
			try 
			{
				sc = new Scanner(file);
				//skips the first line of text
				sc.nextLine();
				while(sc.hasNextLine())
				{
					String str = sc.nextLine();
					String[] line = str.split(" ");
					input.add(Integer.parseInt(line[0]));
					input.add(Integer.parseInt(line[1]));
				}		
			}
			catch (FileNotFoundException e) 
			{
				System.out.println("File not found");
			}
			return input;
		}
	}
}

File: Assignment1/PS.c
/**
 * This class contains the first priority scheduling algorithm
 * @author Eddy Qiang - CPSC 449 - Tutorial T03
 * @version 1.1
 * @since October 14, 2018
 */
#include <stdio.h>
#include <stdlib.h>

int main()
{
	int s = 20;
	char *text1 = (char*)malloc(s*sizeof(char));
	char *text2 = (char*)malloc(s*sizeof(char));
	char *inputFileName = (char*)malloc(s*sizeof(char));
	char *outputFileName = (char*)malloc(s*sizeof(char));
	char *char_b = (char*)malloc(s*sizeof(char));
	char *char_p = (char*)malloc(s*sizeof(char));
	int n;
	int temp;
	int num;
	float totalTime = 0;
	float averageWaitTime = 0;
	//prompt for number of processes
	printf("Enter the number of processes: \n");
	scanf("%d", &num);
	int *burstTime = (int*)malloc(num*sizeof(int));
	int *priority = (int*)malloc(num*sizeof(int));
	int *arr = (int*)malloc(num*sizeof(int));
	int *PS_output = (int*)malloc(num*sizeof(int));
	//prompt for input file name
	printf("Enter the input file name: \n");
	scanf("%s", inputFileName);
	//prompt for output file name
	printf("Enter the output file name: \n");
	scanf("%s", outputFileName);
	//open file for reading
    FILE *input;
    input = fopen(inputFileName, "r");
    free(inputFileName);
    fscanf(input, "%s", text1);
    fscanf(input, "%s", text2);
    free(text1);
    free(text2);

    for(int x = 0; x < num; x++)
	{
    	 fscanf(input, "%s", char_b);
    	 fscanf(input, "%s", char_p);

    	 int btime = atoi(char_b);
    	 int prior = atoi(char_p);

    	 burstTime[x] = btime;
    	 priority[x] = prior;
   	}
    //sort the processes by priority
	for(int i=0;i<num;i++)
	{
		n = i;
		for(int j=i+1;j<num;j++)
		{
			if (priority[j] < priority[n])
			{
				n = j;
			}
		}
	   temp = priority[i]; priority[i] = priority[n]; priority[n] = temp;
	   temp = burstTime[i]; burstTime[i] = burstTime[n]; burstTime[n] = temp;
	   temp = arr[i];  arr[i] = arr[n]; arr[n] = temp;
	}
	PS_output[0] = 0;
	//calculate waiting time
	for(int i=1;i<num;i++)
	{
		PS_output[i] = 0;
		for(int j=0;j<i;j++)
		{
			PS_output[i] += burstTime[j];
		}
		totalTime += PS_output[i];
	}
	FILE *output = fopen(outputFileName, "w");
	free(outputFileName);
	//write wait times
	for(int i=0;i<num;i++)
	{
		fprintf(output, "%d\n", PS_output[i]);
	}
	averageWaitTime = totalTime/num;
	//write average time
	fprintf(output, "%f\n", averageWaitTime);
	free(char_b);
	free(char_p);
	free(burstTime);
	free(priority);
	free(arr);
	free(PS_output);
	return 0;
}
